circleci/s3-upload

```diff
@@ -37,32 +37,36 @@ if [[ $# -eq 4 ]]; then
         exit 1
     fi
     CONTENT_ENCODING=$2
-    SOURCE_DIR=$3
+    SOURCE_DIR_OR_FILE=$3
     S3_BUCKET_URL=$4
 elif [[ $# -eq 2 ]]; then
-    SOURCE_DIR=$1
+    SOURCE_DIR_OR_FILE=$1
     S3_BUCKET_URL=$2
 else
     echo "$USAGE"
     exit 1
 fi
 
-DIR_TO_UPLOAD="$SOURCE_DIR"
+DIR_OR_FILE_TO_UPLOAD="$SOURCE_DIR_OR_FILE"
 ADDITIONAL_S3_FLAGS=""
 
+if [[ -d $SOURCE_DIR_OR_FILE ]]; then
+    ADDITIONAL_S3_FLAGS="--recursive"
+fi
+
 if [[ $CONTENT_ENCODING == "gzip" ]]; then
     rm -rf $WORK_DIR
     mkdir -p $WORK_DIR
-    cp -r $SOURCE_DIR $WORK_DIR
+    cp -r $SOURCE_DIR_OR_FILE $WORK_DIR
     find $WORK_DIR -type f -exec gzip --best {} \; -exec mv {}.gz {} \;
 
-    DIR_TO_UPLOAD="$WORK_DIR/$(basename $SOURCE_DIR)"
-    ADDITIONAL_S3_FLAGS='--content-encoding gzip'
+    DIR_OR_FILE_TO_UPLOAD="$WORK_DIR/$(basename $SOURCE_DIR_OR_FILE)"
+    ADDITIONAL_S3_FLAGS="$ADDITIONAL_S3_FLAGS --content-encoding gzip"
 fi
 
 install_awscli
 
 echo "Uploading files to S3..."
-echo "  Source: $SOURCE_DIR"
+echo "  Source: $SOURCE_DIR_OR_FILE"
 echo "  Desination: $S3_BUCKET_URL"
-aws s3 cp $DIR_TO_UPLOAD $S3_BUCKET_URL --recursive --acl "private" --cache-control "max-age=31536000" $ADDITIONAL_S3_FLAGS
+aws s3 cp $DIR_OR_FILE_TO_UPLOAD $S3_BUCKET_URL --acl "private" --cache-control "max-age=31536000" $ADDITIONAL_S3_FLAGS
```

Comment:
It seems to me like uploading a directory doesn't work without recursive, based on a quick test. Can you double check that your test runs worked?

Reply:
Here's my test
```
quillen:integration-testing-service taylors$ mkdir test
quillen:integration-testing-service taylors$ touch test/{a,b,c}
quillen:integration-testing-service taylors$ ls test/
a	b	c
quillen:integration-testing-service taylors$ aws s3 cp test s3://assets.clever.com-dev
upload failed: test/ to s3://assets.clever.com-dev/ Parameter validation failed:
Invalid length for parameter Key, value: 0, valid range: 1-inf
quillen:integration-testing-service taylors$ aws s3 cp test s3://assets.clever.com-dev/
upload failed: test/ to s3://assets.clever.com-dev/ Parameter validation failed:
Invalid length for parameter Key, value: 0, valid range: 1-inf
quillen:integration-testing-service taylors$ aws s3 cp test s3://assets.clever.com-dev/ --recursive
upload: test/c to s3://assets.clever.com-dev/c
upload: test/a to s3://assets.clever.com-dev/a
upload: test/b to s3://assets.clever.com-dev/b
```

Reply:
Ah yes, so `--recursive` is definitely required for a directory. That's why we have this block above:

https://github.com/Clever/ci-scripts/blob/0de1e3e6d4557ba816f01863db21a49a91618c53/circleci/s3-upload#L53-L55

Reply:
Also
```
quillen:integration-testing-service taylors$ aws s3 cp test s3://assets.clever.com-dev/test
upload failed: test/ to s3://assets.clever.com-dev/test [Errno 21] Is a directory: '/Users/taylors/go/src/github.com/Clever/integration-testing-service/test/'
```

Reply:
Ah excellent, missed that. LGTM

Reply:
And for completeness, here are some of my test runs!

```
arsalan@macbook:ci-scripts$ ./circleci/s3-upload test/ s3://assets.clever.com-dev/test/
AWS cli already installed
Uploading files to S3...
  Source: test/
  Desination: s3://assets.clever.com-dev/test/
upload: test/b.css to s3://assets.clever.com-dev/test/b.css   
upload: test/a.js to s3://assets.clever.com-dev/test/a.js     

arsalan@macbook:ci-scripts$ ./circleci/s3-upload test/a.js s3://assets.clever.com-dev/test/a.js
AWS cli already installed
Uploading files to S3...
  Source: test/a.js
  Desination: s3://assets.clever.com-dev/test/a.js
upload: test/a.js to s3://assets.clever.com-dev/test/a.js
```